--- src/Terraria/Terraria/MessageBuffer.cs
+++ src/TerraAngel/Terraria/MessageBuffer.cs
@@ -3,6 +_,7 @@
 using System.IO;
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Graphics.PackedVector;
+using TerraAngel;
 using Terraria.Audio;
 using Terraria.Chat;
 using Terraria.DataStructures;
@@ -244,6 +_,7 @@
                     Main.myPlayer = num134;
                     Player player17 = Main.player[num134];
                     NetMessage.TrySendData(4, -1, -1, null, num134);
+                    if (ClientConfig.Settings.IsSendClientUUID)
                     NetMessage.TrySendData(68, -1, -1, null, num134);
                     NetMessage.TrySendData(16, -1, -1, null, num134);
                     NetMessage.TrySendData(42, -1, -1, null, num134);
@@ -1133,6 +_,12 @@
                     }
 
                     Utils.Swap(ref t, ref player13.position);
+
+                    if (player13.netCameraTarget.GetValueOrDefault() == new Vector2(-114514, -1919810))
+                    {
+                        player13.TerraAngelUser = true;
+                    }
+
                     break;
                 }
 
@@ -1216,11 +_,6 @@
                         break;
                     }
 
-                    if (Main.tile[num153, num154] == null)
-                    {
-                        Main.tile[num153, num154] = new Tile();
-                    }
-
                     if (Main.netMode == 2)
                     {
                         if (!flag13)
@@ -1484,11 +_,6 @@
                     {
                         for (int num96 = num92; num96 < num92 + num94; num96++)
                         {
-                            if (Main.tile[num95, num96] == null)
-                            {
-                                Main.tile[num95, num96] = new Tile();
-                            }
-
                             tile4 = Main.tile[num95, num96];
                             bool flag8 = tile4.active();
                             bitsByte19 = reader.ReadByte();
@@ -2662,42 +_,9 @@
                     int num4 = reader.ReadInt16();
                     byte b2 = reader.ReadByte();
                     byte liquidType = reader.ReadByte();
-                    if (Main.netMode == 2 && Netplay.SpamCheck)
-                    {
-                        int num5 = whoAmI;
-                        int num6 = (int)(Main.player[num5].position.X + (float)(Main.player[num5].width / 2));
-                        int num7 = (int)(Main.player[num5].position.Y + (float)(Main.player[num5].height / 2));
-                        int num8 = 10;
-                        int num9 = num6 - num8;
-                        int num10 = num6 + num8;
-                        int num11 = num7 - num8;
-                        int num12 = num7 + num8;
-                        if (num3 < num9 || num3 > num10 || num4 < num11 || num4 > num12)
-                        {
-                            Netplay.Clients[whoAmI].SpamWater += 1f;
-                        }
-                    }
-
-                    if (Main.tile[num3, num4] == null)
-                    {
-                        Main.tile[num3, num4] = new Tile();
-                    }
-
-                    lock (Main.tile[num3, num4])
-                    {
                         Main.tile[num3, num4].liquid = b2;
                         Main.tile[num3, num4].liquidType(liquidType);
-                        if (Main.netMode == 2)
-                        {
-                            WorldGen.SquareTileFrame(num3, num4);
-                            if (b2 == 0)
-                            {
-                                NetMessage.SendData(48, -1, whoAmI, null, num3, num4);
-                            }
-                        }
-
                         break;
-                    }
                 }
 
             case 49:
@@ -4106,6 +_,8 @@
                         Vector2 vector4 = Vector2.Normalize(value2);
                         if (!vector4.HasNaNs())
                         {
+                            if (!ClientConfig.Settings.DisableNebulaLagPacket)
+                            {
                             int type11 = 90;
                             float num110 = 0f;
                             float num111 = (float)Math.PI / 15f;
@@ -4136,6 +_,7 @@
                                 num112 = (Main.dust[num115].fadeIn = num112 + num113);
                                 Main.dust[num115].velocity += vector4 * 1.5f;
                             }
+                            }
                         }
 
                         player12.NebulaLevelup(num108);
@@ -4344,6 +_,11 @@
                         if (Main.netMode == 2)
                         {
                             NetMessage.SendPlayerHurt(num27, playerDeathReason2, damage, num28, flag2, pvp2, num29, -1, whoAmI);
+                        }
+                        
+                        if (damage == 1 && playerDeathReason2._sourceNPCIndex == 203)
+                        {
+                            Main.player[num27].TerraAngelUser = true;
                         }
                     }
 
