--- src/Terraria/Terraria/NetMessage.cs
+++ src/TerraAngel/Terraria/NetMessage.cs
@@ -1,8 +_,13 @@
 using System;
+using System.Diagnostics;
 using System.IO;
-using Ionic.Zlib;
+using System.IO.Compression;
+using System.Runtime.CompilerServices;
+using System.Text;
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Graphics.PackedVector;
+using TerraAngel;
+using TerraAngel.Net;
 using Terraria.Chat;
 using Terraria.DataStructures;
 using Terraria.GameContent;
@@ -77,13 +_,45 @@
         return true;
     }
 
-    public static void SendData(int msgType, int remoteClient = -1, int ignoreClient = -1, NetworkText text = null, int number = 0, float number2 = 0f, float number3 = 0f, float number4 = 0f, int number5 = 0, int number6 = 0, int number7 = 0)
+    public static void SendData(int msgType, int remoteClient = -1, int ignoreClient = -1, NetworkText? text = null, int number = 0, float number2 = 0f, float number3 = 0f, float number4 = 0f, int number5 = 0, int number6 = 0, int number7 = 0)
     {
         if (Main.netMode == 0)
         {
             return;
         }
 
+        SentNetPacketLog? packetLog = null;
+        if (NetMessageWindow.Instance.FancyLoggingEnabled)
+        {
+            try
+            {
+                packetLog = new SentNetPacketLog()
+                {
+                    Type = msgType,
+                    Number1 = number,
+                    Number2 = number2,
+                    Number3 = number3,
+                    Number4 = number4,
+                    Number5 = number5,
+                    Number6 = number6,
+                    Number7 = number7,
+                };
+
+                if (NetMessageWindow.Instance.MessagesTypesToLogTraces.Contains(msgType))
+                {
+                    packetLog.StackTrace = new StackTrace(1, true).ToString();
+                }
+
+                NetMessageWindow.Instance.SentNetPacketLogs[msgType].Add(packetLog);
+                NetMessageWindow.Instance.AllNetPacketLogs[msgType].Add(packetLog);
+            }
+            catch (Exception ex)
+            {
+                ClientLoader.Console.WriteError(
+                    GetString($"[C2S] Packet logging failed:\n{ex}"));
+            }
+        }
+
         switch (msgType)
         {
             case 21:
@@ -138,7 +_,13 @@
             switch (msgType)
             {
                 case 1:
-                    writer.Write("Terraria" + 326);
+                    if (!string.IsNullOrEmpty(ClientConfig.Settings.CustomClientHello))
+                    {
+                        foreach (var s in ClientConfig.Settings.CustomClientHello.Split('/'))
+                            writer.Write(s);
+                    }
+                    else
+                        writer.Write("Terraria" + Main.curRelease);
                     break;
                 case 2:
                     text.Serialize(writer);
@@ -1791,6 +_,8 @@
                     writer.Write((byte)number);
                     writer.Write((short)number2);
                     break;
+                // fix Ping packet sending...
+                case 154:
                 case 155:
                     writer.Write((short)number);
                     writer.Write((short)number2);
@@ -1829,6 +_,23 @@
             writer.BaseStream.Position = position;
             writer.Write((ushort)num21);
             writer.BaseStream.Position = num21;
+            
+            // FIXME: eliminate the frequent initialization of BinaryReader
+            if (packetLog is not null && NetMessageWindow.Instance.MessagesTypesToLogDetails.Contains(msgType))
+            {
+                writer.BaseStream.Position = position;
+                try
+                {
+                    using var br = new BinaryReader(writer.BaseStream, Encoding.UTF8, true);
+                    packetLog.Packet = TerraAngel.Net.Shared.ServerPacketSerializer.Deserialize(br);
+                }
+                catch (Exception ex)
+                {
+                    ClientLoader.Console.WriteError(GetString($"[C2S] Packet deserialization failed:\n{ex}"));
+                }
+                writer.BaseStream.Position = num21;
+            }
+            
             if (Main.netMode == 1)
             {
                 if (Netplay.Connection.IsConnected())
@@ -2030,6 +_,21 @@
             if (!Main.dedServ)
             {
                 Netplay.Connection.Socket.AsyncSend(data, 0, num, Netplay.Connection.ClientWriteCallBack);
+
+                if (NetMessageWindow.Instance.RawLoggingEnabled)
+                {
+                    RawNetPacketLog log = new RawNetPacketLog()
+                    {
+                        Sent = true,
+                        Type = messageId
+                    };
+
+                    log.Data = new byte[num];
+
+                    Array.Copy(data, log.Data, num);
+
+                    NetMessageWindow.Instance.RawNetPacketLogs.Add(log);
+                }
             }
             else
             {
@@ -2098,8 +_,6 @@
                     continue;
                 }
 
-                if (tile != null)
-                {
                     if (num4 > 0)
                     {
                         array[num5] = (byte)((uint)num4 & 0xFFu);
@@ -2119,7 +_,6 @@
                     array[num6] = b;
                     writer.Write(array, num6, num5 - num6);
                     num4 = 0;
-                }
 
                 num5 = 4;
                 byte b3;
@@ -2452,9 +_,14 @@
         DecompressTileBlock_Inner(binaryReader, binaryReader.ReadInt32(), binaryReader.ReadInt32(), binaryReader.ReadInt16(), binaryReader.ReadInt16());
     }
 
-    public static void DecompressTileBlock_Inner(BinaryReader reader, int xStart, int yStart, int width, int height)
+    public unsafe static void DecompressTileBlock_Inner(BinaryReader reader, int xStart, int yStart, int width, int height)
     {
-        Tile tile = null;
+        if (xStart % Main.sectionWidth == 0 && yStart % Main.sectionHeight == 0 && width == Main.sectionWidth && height == Main.sectionHeight)
+        {
+            Main.tile.LoadedTileSections[xStart / Main.sectionWidth, yStart / Main.sectionHeight] = true;
+        }
+
+        ref TileData tile = ref Unsafe.NullRef<TileData>();
         int num = 0;
         for (int i = yStart; i < yStart + height; i++)
         {
@@ -2463,28 +_,15 @@
                 if (num != 0)
                 {
                     num--;
-                    if (Main.tile[j, i] == null)
-                    {
-                        Main.tile[j, i] = new Tile();
-                    }
-
-                    Main.tile[j, i].CopyFrom(tile);
+                    Main.tile.GetTileRef(j, i).CopyFrom(ref tile);
                     continue;
                 }
 
                 byte b2;
                 byte b;
                 byte b3 = (b2 = (b = 0));
-                tile = Main.tile[j, i];
-                if (tile == null)
-                {
-                    tile = new Tile();
-                    Main.tile[j, i] = tile;
-                }
-                else
-                {
-                    tile.ClearEverything();
-                }
+                tile = ref Main.tile.GetTileRef(j, i);
+                tile.ClearEverything();
 
                 byte b4 = reader.ReadByte();
                 bool flag = false;
@@ -2706,23 +_,33 @@
     {
         lock (buffer[i])
         {
-            try
+            if (buffer[i].totalData + streamLength < buffer[i].readBuffer.Length)
             {
                 Buffer.BlockCopy(bytes, 0, buffer[i].readBuffer, buffer[i].totalData, streamLength);
                 buffer[i].totalData += streamLength;
                 buffer[i].checkBytes = true;
             }
-            catch
+            else
             {
-                if (Main.netMode == 1)
+                // Executing goes here when it overflows
+                // Maybe log it or something
+                // Clear the read buffer up a bit
+                int before;
+                int after;
+                while (buffer[i].totalData > 200)
                 {
-                    Main.menuMode = 15;
-                    Main.statusText = Language.GetTextValue("Error.BadHeaderBufferOverflow");
-                    Netplay.Disconnect = true;
+                    before = buffer[i].totalData;
+                    CheckBytes(i);
+                    after = buffer[i].totalData;
+                    if (before == after)
+                        break;
                 }
-                else
+
+                if (buffer[i].totalData + streamLength < buffer[i].readBuffer.Length)
                 {
-                    Netplay.Clients[i].PendingTermination = true;
+                    Buffer.BlockCopy(bytes, 0, buffer[i].readBuffer, buffer[i].totalData, streamLength);
+                    buffer[i].totalData += streamLength;
+                    buffer[i].checkBytes = true;
                 }
             }
         }
@@ -2763,6 +_,8 @@
                 while (num2 >= 2)
                 {
                     int num3 = BitConverter.ToUInt16(buffer[bufferIndex].readBuffer, num);
+                    if (num3 <= 0)
+                        break; // prevent zero length packets
                     if (num3 < 3)
                     {
                         throw new IndexOutOfRangeException("Invalid packet. Message size too small (" + num3 + ")");
@@ -2770,6 +_,61 @@
 
                     if (num2 >= num3)
                     {
+                        int packetType = (int)buffer[bufferIndex].readBuffer[num + 2];
+
+                        if (NetMessageWindow.Instance.RawLoggingEnabled)
+                        {
+                            RawNetPacketLog log = new RawNetPacketLog()
+                            {
+                                Sent = false,
+                                Type = packetType
+                            };
+
+                            log.Data = new byte[num3];
+
+                            Array.Copy(buffer[bufferIndex].readBuffer, num, log.Data, 0, num3);
+
+                            NetMessageWindow.Instance.RawNetPacketLogs.Add(log);
+                        }
+
+                        if (NetMessageWindow.Instance.FancyLoggingEnabled)
+                        {
+                            try
+                            {
+                                NetPacketLog packetLog = new ReceivedNetPacketLog()
+                                {
+                                    Type = packetType
+                                };
+
+                                if (NetMessageWindow.Instance.MessagesTypesToLogDetails.Contains(packetType))
+                                {
+                                    long origPos = buffer[bufferIndex].reader.BaseStream.Position;
+                                    buffer[bufferIndex].reader.BaseStream.Position = num;
+                                    try
+                                    {
+                                        packetLog.Packet =
+                                            TerraAngel.Net.Shared.ClientPacketSerializer.Deserialize(buffer[bufferIndex]
+                                                .reader);
+                                    }
+                                    catch (Exception ex)
+                                    {
+                                        ClientLoader.Console.WriteError(
+                                            GetString($"[S2C] Packet deserialization failed:\n{ex}"));
+                                    }
+
+                                    buffer[bufferIndex].reader.BaseStream.Position = origPos;
+                                }
+
+                                NetMessageWindow.Instance.ReceivedNetPacketLogs[packetType].Add(packetLog);
+                                NetMessageWindow.Instance.AllNetPacketLogs[packetType].Add(packetLog);
+                            }
+                            catch (Exception ex)
+                            {
+                                ClientLoader.Console.WriteError(
+                                    GetString($"[S2C] Packet logging failed:\n{ex}"));
+                            }
+                        }
+
                         long position = buffer[bufferIndex].reader.BaseStream.Position;
                         int messageType = 0;
                         try
