--- src/Terraria/Terraria/Player.cs
+++ src/TerraAngel/Terraria/Player.cs
@@ -39,6 +_,7 @@
 using Terraria.UI.Gamepad;
 using Terraria.Utilities;
 using Terraria.WorldBuilding;
+using TerraAngel.Tools;
 
 namespace Terraria;
 public class Player : Entity, IFixLoadedData
@@ -751,6 +_,8 @@
     public bool active;
     public bool host;
     public Minecart.Customization MinecartSettings;
+
+    public bool TerraAngelUser = false;
     public int emoteTime;
     public CreativeUnlocksTracker creativeTracker;
     private static byte[] ENCRYPTION_KEY = new UnicodeEncoding().GetBytes("h3y_gUyZ");
@@ -1184,8 +_,10 @@
     public int manaPotionDelay;
     public byte difficulty;
     public byte wetSlime;
-    public HitTile hitTile;
+    public HitTile? _hitTile;
-    public HitTile hitReplace;
+    public HitTile? _hitReplace;
+    public HitTile hitTile { get { if (_hitTile is null) _hitTile = new HitTile(); return _hitTile; } }
+    public HitTile hitReplace { get { if (_hitReplace is null) _hitReplace = new HitTile(); return _hitReplace; } }
     public int jump;
     public int head = -1;
     public int body = -1;
@@ -12017,6 +_,7 @@
 
     public void Yoraiz0rEye()
     {
+        if ((OptimizationTool.OptimizationToolCache?.DisableDust ?? false) || ClientConfig.Settings.IgnoreReLogicBullshit) return;
         int num = 0;
         num += bodyFrame.Y / 56;
         if (num >= Main.OffsetsPlayerHeadgear.Length)
@@ -14657,7 +_,7 @@
             shinyStone = true;
         }
 
-        if (currentItem.type == 4989)
+        if (currentItem.type == 4989 || ToolManager.GetTool<InfiniteFlightTool>().Enabled)
         {
             empressBrooch = true;
             moveSpeed += 0.075f;
@@ -18028,7 +_,7 @@
                 }
 
                 int num5 = tile.frameY / 22;
-                if (num5 < TorchID.Count)
+                if (num5 < TorchID.Count && num5 >= 0)
                 {
                     nearbyTorch[num5] = true;
                     if (num5 == 17 && (tile.liquid == 0 || tile.liquidType() != 0))
@@ -18570,7 +_,9 @@
         return (int)((float)damage * (1f + num3));
     }
 
+    public void ResetEffects() => TerraAngel.Hooks.PlayerHooks.PlayerResetEffectsHook((x) => x._ResetEffects(), this);
+
-    public void ResetEffects()
+    private void _ResetEffects()
     {
         IsAllowedToHoldItems = true;
         if (rapidAttackBonus > 0f)
@@ -20074,16 +_,6 @@
         }
 
         float num3 = position.Y;
-        if (Main.tile[num, num2 - 1] == null)
-        {
-            Main.tile[num, num2 - 1] = new Tile();
-        }
-
-        if (Main.tile[num, num2 + 1] == null)
-        {
-            Main.tile[num, num2 + 1] = new Tile();
-        }
-
         if ((!Main.tile[num, num2 - 1].active() || !Main.tileRope[Main.tile[num, num2 - 1].type]) && (!Main.tile[num, num2 + 1].active() || !Main.tileRope[Main.tile[num, num2 + 1].type]))
         {
             num3 = num2 * 16 + 22;
@@ -24065,10 +_,6 @@
                 int num3 = (int)((position.Y + (float)num2 - 44f * gravDir) / 16f);
                 if (Main.tile[num, num3] != null && Main.tile[num, num3].liquid < 128)
                 {
-                    if (Main.tile[num, num3] == null)
-                    {
-                        Main.tile[num, num3] = new Tile();
-                    }
 
                     if (!Main.tile[num, num3].active() || !Main.tileSolid[Main.tile[num, num3].type] || Main.tileSolidTop[Main.tile[num, num3].type])
                     {
@@ -24675,21 +_,6 @@
     public static Tile GetFloorTile(int x, int y)
     {
         Tile result = null;
-        if (Main.tile[x - 1, y] == null)
-        {
-            Main.tile[x - 1, y] = new Tile();
-        }
-
-        if (Main.tile[x + 1, y] == null)
-        {
-            Main.tile[x + 1, y] = new Tile();
-        }
-
-        if (Main.tile[x, y] == null)
-        {
-            Main.tile[x, y] = new Tile();
-        }
-
         if (Main.tile[x, y].nactive() && Main.tileSolid[Main.tile[x, y].type])
         {
             result = Main.tile[x, y];
@@ -24708,7 +_,7 @@
 
     public static int GetFloorTileType(int x, int y)
     {
-        return ((int?)GetFloorTile(x, y)?.type) ?? (-1);
+        return (int)GetFloorTile(x, y).type;
     }
 
     private void MakeFloorDust(bool Falling, int type, int frameX, int frameY, int paintColor)
@@ -25375,6 +_,11 @@
             LockGamepadTileInteractions();
         }
 
+        if (ToolManager.GetTool<InstantRespawnTool>().Enabled)
+        {
+            respawnTimer = 0;
+        }
+
         maxFallSpeed = 10f;
         gravity = defaultGravity;
         jumpHeight = 15;
@@ -26565,21 +_,6 @@
             tileTargetY = 5;
         }
 
-        if (Main.tile[tileTargetX - 1, tileTargetY] == null)
-        {
-            Main.tile[tileTargetX - 1, tileTargetY] = new Tile();
-        }
-
-        if (Main.tile[tileTargetX + 1, tileTargetY] == null)
-        {
-            Main.tile[tileTargetX + 1, tileTargetY] = new Tile();
-        }
-
-        if (Main.tile[tileTargetX, tileTargetY] == null)
-        {
-            Main.tile[tileTargetX, tileTargetY] = new Tile();
-        }
-
         if (inventory[selectedItem].axe > 0 && !Main.tile[tileTargetX, tileTargetY].active() && inventory[selectedItem].createWall <= 0 && (inventory[selectedItem].hammer <= 0 || inventory[selectedItem].axe != 0))
         {
             if (Main.tile[tileTargetX - 1, tileTargetY].active() && Main.tile[tileTargetX - 1, tileTargetY].type == 323)
@@ -27481,16 +_,6 @@
                     int num35 = num34 * 16 + 8 - width / 2;
                     float y2 = position.Y;
                     y2 = num32 * 16 + 22;
-                    if (Main.tile[num34, num32 - 1] == null)
-                    {
-                        Main.tile[num34, num32 - 1] = new Tile();
-                    }
-
-                    if (Main.tile[num34, num32 + 1] == null)
-                    {
-                        Main.tile[num34, num32 + 1] = new Tile();
-                    }
-
                     if (WorldGen.IsRope(num34, num32 - 1) || WorldGen.IsRope(num34, num32 + 1))
                     {
                         y2 = num32 * 16 + 22;
@@ -27534,11 +_,6 @@
                 pulley = false;
             }
 
-            if (Main.tile[num31, num32] == null)
-            {
-                Main.tile[num31, num32] = new Tile();
-            }
-
             if (!WorldGen.IsRope(num31, num32))
             {
                 pulley = false;
@@ -27601,11 +_,6 @@
                 flag16 = true;
             }
 
-            if (Main.tile[num36, num37] == null)
-            {
-                Main.tile[num36, num37] = new Tile();
-            }
-
             if (!WorldGen.IsRope(num36, num37))
             {
                 flag15 = false;
@@ -29536,35 +_,45 @@
             CheckCrackedBrickBreak();
         }
 
+        if ((!ToolManager.GetTool<NoClipTool>().Enabled && i == Main.myPlayer) || (i != Main.myPlayer && Main.tile.IsTileInLoadedSection(((int)(position.X / 16f)), ((int)(position.Y / 16f)))))
+        {
         if (!shimmering)
         {
             SlopeDownMovement();
+            // patch context
             bool flag28 = mount.Type == 7 || mount.Type == 8 || mount.Type == 12 || mount.Type == 44 || mount.Type == 49;
             bool flag29 = mount.Type == 48;
+            // patch context
             if (!mount.Active)
             {
                 flag28 = false;
                 flag29 = false;
             }
 
+            // patch context
             if (velocity.Y == gravity && !IsRidingTracks && !flag29 && !flag28)
             {
                 Collision.StepDown(ref position, ref velocity, width, height, ref stepSpeed, ref gfxOffY, (int)gravDir, waterWalk || waterWalk2);
             }
 
+            // patch context
             if (gravDir == -1f)
             {
+                // patch context
                 if ((carpetFrame != -1 || velocity.Y <= gravity) && !controlUp)
                 {
                     Collision.StepUp(ref position, ref velocity, width, height, ref stepSpeed, ref gfxOffY, (int)gravDir, controlUp);
                 }
+                // patch context
             }
             else if ((carpetFrame != -1 || velocity.Y >= gravity) && !controlDown && !IsRidingTracks && !flag28 && grappling[0] == -1)
             {
                 Collision.StepUp(ref position, ref velocity, width, height, ref stepSpeed, ref gfxOffY, (int)gravDir, controlUp);
             }
+            // patch context
         }
 
+        // patch context
         oldPosition = position;
         oldDirection = direction;
         bool falling = false;
@@ -29789,20 +_,22 @@
         {
             FloorVisuals(falling);
         }
-
+        // patch context
         if (whoAmI == Main.myPlayer && !shimmering)
         {
             Collision.SwitchTiles(this, position, width, height, oldPosition, 5);
         }
-
+        // patch context
         PressurePlateHelper.UpdatePlayerPosition(this);
         BordersMovement();
         numMinions = 0;
         slotsMinions = 0f;
+        // patch context
         if (Main.netMode != 2 && mount.Type != 8)
         {
             ItemCheck_ManageRightClickFeatures();
         }
+        }
 
         ItemCheckWrapped(i);
         PlayerFrame();
@@ -33649,7 +_,9 @@
         }
     }
 
+    private void ItemCheckWrapped(int i) => TerraAngel.Hooks.PlayerHooks.PlayerItemCheckHook((x, y) => x._ItemCheckWrapped(y), this, i);
+
-    private void ItemCheckWrapped(int i)
+    private void _ItemCheckWrapped(int i)
     {
         int num;
         if (!controlLeft && !controlRight && !controlUp && !controlDown && PlayerInput.UsingGamepad && Main.SmartCursorIsUsed)
@@ -33903,11 +_,6 @@
 
     public void TileInteractionsCheck(int myX, int myY)
     {
-        if (Main.tile[myX, myY] == null)
-        {
-            Main.tile[myX, myY] = new Tile();
-        }
-
         if (Main.tile[myX, myY].active())
         {
             TileInteractionsMouseOver(myX, myY);
@@ -35307,14 +_,16 @@
         if (Recipe.TileUsedInRecipes[type])
         {
             int baseOption = 0;
-            MapHelper.GetTileBaseOption(myX, myY, type, Main.tile[myX, myY], ref baseOption);
+            ref TileData tile = ref Main.tile.GetTileRef(myX, myY);
+            MapHelper.GetTileBaseOption(myX, myY, type, ref tile, ref baseOption);
             return new NewCraftingUI.CraftStationRecipeFilter(type, baseOption);
         }
 
         if (TileID.Sets.CountsAsWaterForCrafting[type])
         {
             int baseOption2 = 0;
-            MapHelper.GetTileBaseOption(myX, myY, type, Main.tile[myX, myY], ref baseOption2);
+            ref TileData tile = ref Main.tile.GetTileRef(myX, myY);
+            MapHelper.GetTileBaseOption(myX, myY, type, ref tile, ref baseOption2);
             return new NewCraftingUI.WaterSourceRecipeFilter(type, baseOption2);
         }
 
@@ -39826,6 +_,17 @@
 
     public void Teleport(Vector2 newPos, int Style = 0, int extraInfo = 0)
     {
+
+        if (ToolManager.GetTool<TeleportControlTool>().LogTargetPosition)
+        {
+            ClientLoader.Console.WriteLine(GetString($"Teleported to {newPos.ToString()} (Style: {Style}, ExtraInfo: {extraInfo})"));
+        }
+
+        if (!ToolManager.GetTool<TeleportControlTool>().AllowTeleport)
+        {
+            return;
+        }
+
         try
         {
             _funkytownAchievementCheckCooldown = 100;
@@ -40673,6 +_,17 @@
 
     public double Hurt(PlayerDeathReason damageSource, int Damage, int hitDirection, bool pvp = false, bool quiet = false, bool Crit = false, int cooldownCounter = -1, bool dodgeable = true)
     {
+        if (whoAmI == Main.myPlayer && ToolManager.GetTool<AntiHurtTool>().Enabled)
+        {
+            statLife = statLifeMax2;
+            if (ToolManager.GetTool<AntiHurtTool>().FramesSinceLastLifePacket == 0)
+            {
+                ToolManager.GetTool<AntiHurtTool>().FramesSinceLastLifePacket = 6;
+                NetMessage.SendData(MessageID.PlayerLife, -1, -1, null, whoAmI);
+            }
+            return 0.0d;
+        }
+
         if (shimmering && AllowShimmerDodge(damageSource, cooldownCounter, dodgeable))
         {
             return 0.0;
@@ -41367,6 +_,12 @@
 
     public void KillMe(PlayerDeathReason damageSource, double dmg, int hitDirection, bool pvp = false)
     {
+        if (whoAmI == Main.myPlayer && ToolManager.GetTool<AntiHurtTool>().Enabled)
+        {
+            NetMessage.SendData(MessageID.PlayerLife, -1, -1, null, whoAmI);
+            return;
+        }
+
         if (creativeGodMode || dead)
         {
             return;
@@ -42280,13 +_,23 @@
 
     private void PlaceThing_Walls()
     {
-        if (inventory[selectedItem].createWall < 0 || !IsInTileInteractionRange(tileTargetX, tileTargetY, TileReachCheckSettings.Simple, inventory[selectedItem].tileBoost + blockRange))
+        bool blockedByRangeCheck = !IsInTileInteractionRange(tileTargetX, tileTargetY, TileReachCheckSettings.Simple, inventory[selectedItem].tileBoost + blockRange);
+
+        if (ToolManager.GetTool<InfiniteReachTool>().Enabled || BuilderModeTool.Enabled)
+            blockedByRangeCheck = false;
+
+        if (inventory[selectedItem].createWall < 0 || blockedByRangeCheck)
         {
             return;
         }
 
+        bool wallPlaceBlockedByNoFreeSpace = (!Main.tile[tileTargetX + 1, tileTargetY].active() && Main.tile[tileTargetX + 1, tileTargetY].wall <= 0 && !Main.tile[tileTargetX - 1, tileTargetY].active() && Main.tile[tileTargetX - 1, tileTargetY].wall <= 0 && !Main.tile[tileTargetX, tileTargetY + 1].active() && Main.tile[tileTargetX, tileTargetY + 1].wall <= 0 && !Main.tile[tileTargetX, tileTargetY - 1].active() && Main.tile[tileTargetX, tileTargetY - 1].wall <= 0);
+
+        if (BuilderModeTool.Enabled)
+            wallPlaceBlockedByNoFreeSpace = false;
+
         cursorItemIconEnabled = true;
-        if (!ItemTimeIsZero || itemAnimation <= 0 || !controlUseItem || (!Main.tile[tileTargetX + 1, tileTargetY].active() && Main.tile[tileTargetX + 1, tileTargetY].wall <= 0 && !Main.tile[tileTargetX - 1, tileTargetY].active() && Main.tile[tileTargetX - 1, tileTargetY].wall <= 0 && !Main.tile[tileTargetX, tileTargetY + 1].active() && Main.tile[tileTargetX, tileTargetY + 1].wall <= 0 && !Main.tile[tileTargetX, tileTargetY - 1].active() && Main.tile[tileTargetX, tileTargetY - 1].wall <= 0) || Main.tile[tileTargetX, tileTargetY].wall == inventory[selectedItem].createWall)
+        if (((!ItemTimeIsZero || itemAnimation <= 0) && !BuilderModeTool.Enabled) || !controlUseItem || wallPlaceBlockedByNoFreeSpace || Main.tile[tileTargetX, tileTargetY].wall == inventory[selectedItem].createWall)
         {
             return;
         }
@@ -42418,7 +_,13 @@
     {
         Item item = inventory[selectedItem];
         int tileToCreate = item.createTile;
-        if (tileToCreate < 0 || !IsInTileInteractionRange(tileTargetX, tileTargetY, TileReachCheckSettings.Simple, item.tileBoost + blockRange) || !SmartCursorHelper.TileTargetDesired())
+
+        bool blockedByRangeCheck = !IsInTileInteractionRange(tileTargetX, tileTargetY, TileReachCheckSettings.Simple, item.tileBoost + blockRange);
+
+        if (ToolManager.GetTool<InfiniteReachTool>().Enabled || BuilderModeTool.Enabled)
+            blockedByRangeCheck = false;
+
+        if (tileToCreate < 0 || blockedByRangeCheck || !SmartCursorHelper.TileTargetDesired())
         {
             return;
         }
@@ -42448,7 +_,7 @@
             }
         }
 
-        if (canUse && ((!tile.active() && !PlaceThing_Tiles_IsBlockedByLava(item.createTile, item.placeStyle, tile)) || PlaceThing_IsReplaceableBlock(tile) || tileToCreate == 199 || tileToCreate == 23 || tileToCreate == 662 || tileToCreate == 661 || tileToCreate == 2 || tileToCreate == 109 || tileToCreate == 60 || tileToCreate == 70 || tileToCreate == 633 || Main.tileMoss[tileToCreate]) && ItemTimeIsZero && itemAnimation > 0 && controlUseItem)
+        if (canUse && ((!tile.active() && !PlaceThing_Tiles_IsBlockedByLava(item.createTile, item.placeStyle, tile)) || PlaceThing_IsReplaceableBlock(tile) || tileToCreate == 199 || tileToCreate == 23 || tileToCreate == 662 || tileToCreate == 661 || tileToCreate == 2 || tileToCreate == 109 || tileToCreate == 60 || tileToCreate == 70 || tileToCreate == 633 || Main.tileMoss[tileToCreate]) && (ItemTimeIsZero && itemAnimation > 0 || BuilderModeTool.Enabled) && controlUseItem)
         {
             bool? overrideCanPlace = null;
             int? forcedRandom = null;
@@ -42491,6 +_,9 @@
             canPlace = PlaceThing_Tiles_BlockPlacementForAssortedThings(canPlace);
         }
 
+        if (BuilderModeTool.Enabled)
+            canPlace = true;
+
         if (canPlace)
         {
             PlaceThing_Tiles_PlaceIt(newObjectType, data, tileToCreate);
@@ -44688,7 +_,12 @@
 
     private void PlaceThing_PaintScraper()
     {
-        if (ItemID.Sets.IsPaintScraper[inventory[selectedItem].type] && IsInTileInteractionRange(tileTargetX, tileTargetY, TileReachCheckSettings.Simple, inventory[selectedItem].tileBoost + blockRange))
+        bool allowedByRangeCheck = IsInTileInteractionRange(tileTargetX, tileTargetY, TileReachCheckSettings.Simple, inventory[selectedItem].tileBoost + blockRange);
+
+        if (ToolManager.GetTool<InfiniteReachTool>().Enabled || BuilderModeTool.Enabled)
+            allowedByRangeCheck = true;
+
+        if (ItemID.Sets.IsPaintScraper[inventory[selectedItem].type] && allowedByRangeCheck)
         {
             int num = tileTargetX;
             int num2 = tileTargetY;
@@ -44708,7 +_,7 @@
         }
 
         cursorItemIconEnabled = true;
-        if (!ItemTimeIsZero || itemAnimation <= 0 || !controlUseItem)
+        if (((!ItemTimeIsZero || itemAnimation <= 0) && !BuilderModeTool.Enabled) || !controlUseItem)
         {
             return;
         }
@@ -44762,7 +_,7 @@
         }
 
         cursorItemIconEnabled = true;
-        if (ItemTimeIsZero && itemAnimation > 0 && controlUseItem)
+        if ((BuilderModeTool.Enabled || ItemTimeIsZero && itemAnimation > 0) && controlUseItem)
         {
             if (WorldGen.paintTile(x, y, 0, broadCast: true) || WorldGen.paintCoatTile(x, y, 0, broadcast: true))
             {
@@ -44777,7 +_,12 @@
 
     private void PlaceThing_PaintRoller()
     {
-        if ((inventory[selectedItem].type != 1072 && inventory[selectedItem].type != 1544) || !IsInTileInteractionRange(tileTargetX, tileTargetY, TileReachCheckSettings.Simple, inventory[selectedItem].tileBoost + blockRange))
+        bool blockedByRangeCheck = !IsInTileInteractionRange(tileTargetX, tileTargetY, TileReachCheckSettings.Simple, inventory[selectedItem].tileBoost + blockRange);
+
+        if (ToolManager.GetTool<InfiniteReachTool>().Enabled || BuilderModeTool.Enabled)
+            blockedByRangeCheck = false;
+
+        if ((inventory[selectedItem].type != 1072 && inventory[selectedItem].type != 1544) || blockedByRangeCheck)
         {
             return;
         }
@@ -44787,7 +_,7 @@
         if (Main.tile[num, num2] != null && Main.tile[num, num2].wall > 0)
         {
             cursorItemIconEnabled = true;
-            if (ItemTimeIsZero && itemAnimation > 0 && controlUseItem)
+            if ((BuilderModeTool.Enabled || ItemTimeIsZero && itemAnimation > 0) && controlUseItem)
             {
                 TryPainting(num, num2, paintingAWall: true);
             }
@@ -44796,7 +_,12 @@
 
     private void PlaceThing_Paintbrush()
     {
-        if ((inventory[selectedItem].type != 1071 && inventory[selectedItem].type != 1543) || !IsInTileInteractionRange(tileTargetX, tileTargetY, TileReachCheckSettings.Simple, inventory[selectedItem].tileBoost + blockRange))
+        bool blockedByRangeCheck = !IsInTileInteractionRange(tileTargetX, tileTargetY, TileReachCheckSettings.Simple, inventory[selectedItem].tileBoost + blockRange);
+
+        if (ToolManager.GetTool<InfiniteReachTool>().Enabled || BuilderModeTool.Enabled)
+            blockedByRangeCheck = false;
+
+        if ((inventory[selectedItem].type != 1071 && inventory[selectedItem].type != 1543) || blockedByRangeCheck)
         {
             return;
         }
@@ -44806,7 +_,7 @@
         if (Main.tile[num, num2] != null && Main.tile[num, num2].active())
         {
             cursorItemIconEnabled = true;
-            if (ItemTimeIsZero && itemAnimation > 0 && controlUseItem)
+            if ((BuilderModeTool.Enabled || ItemTimeIsZero && itemAnimation > 0) && controlUseItem)
             {
                 TryPainting(num, num2);
             }
@@ -45179,7 +_,7 @@
         return num;
     }
 
-    private void Fishing_GetBait(out int baitPower, out int baitType)
+    public void Fishing_GetBait(out int baitPower, out int baitType)
     {
         baitPower = 0;
         baitType = 0;
@@ -46292,6 +_,7 @@
                 lastVisualizedSelectedItem = item.Clone();
             }
 
+            bool noConsumeInBuilderMode = BuilderModeTool.Enabled && (item.createTile >= 0 || item.createWall >= 0);
             if (whoAmI == Main.myPlayer)
             {
                 if (!dontConsumeWand && itemTimeMax != 0 && itemTime == itemTimeMax && item.tileWand > 0)
@@ -46301,6 +_,7 @@
                     {
                         if (tileWand == inventory[num18].type && inventory[num18].stack > 0)
                         {
+                            if (noConsumeInBuilderMode) break;
                             inventory[num18].stack--;
                             if (inventory[num18].stack <= 0)
                             {
@@ -46356,7 +_,7 @@
 
                     if (flag7)
                     {
-                        if (item.stack > 0)
+                        if (item.stack > 0 && !noConsumeInBuilderMode)
                         {
                             item.stack--;
                         }
@@ -54448,11 +_,6 @@
                 {
                     int num7 = (int)(itemLocation.X + (float)heldItemFrame.Width * 0.8f * (float)direction) / 16;
                     int num8 = (int)(itemLocation.Y + mountOffset + (float)(heldItemFrame.Height / 2)) / 16;
-                    if (Main.tile[num7, num8] == null)
-                    {
-                        Main.tile[num7, num8] = new Tile();
-                    }
-
                     if (Main.tile[num7, num8].active() && TileID.Sets.Campfires[Main.tile[num7, num8].type] && Main.tile[num7, num8].frameY < 54)
                     {
                         miscTimer++;
@@ -55937,6 +_,7 @@
             {
                 if (Main.projectile[i].active && Main.projectile[i].owner == whoAmI && Main.projectile[i].minion && (Main.projectile[i].type == 1093 || Main.projectile[i].type == 1112))
                 {
+                    if (!TerraAngel.Tools.ToolManager.GetTool<InfiniteMinionTool>().Enabled)
                     Main.projectile[i].Kill();
                 }
             }
@@ -60855,8 +_,6 @@
         usedGummyWorm = false;
         usedAmbrosia = false;
         Array.Clear(adjTile, 0, adjTile.Length);
-        hitTile = new HitTile();
-        hitReplace = new HitTile();
         mount = new Mount();
         talkNPC = -1;
         piggyBankProjTracker.Clear();
